File: DSA/LakshaKundra_590013237/Day_89_Question_3.py
def kosaraju(vertices, edges):
    graph = [[] for _ in range(vertices)]
    rev_graph = [[] for _ in range(vertices)]
    for u, v in edges:
        graph[u].append(v)
        rev_graph[v].append(u)
    visited = [False]*vertices
    order = []
    def dfs(u):
        visited[u] = True
        for v in graph[u]:
            if not visited[v]:
                dfs(v)
        order.append(u)
    for i in range(vertices):
        if not visited[i]:
            dfs(i)
    visited = [False]*vertices
    def rev_dfs(u):
        visited[u] = True
        for v in rev_graph[u]:
            if not visited[v]:
                rev_dfs(v)
    scc_count = 0
    for u in reversed(order):
        if not visited[u]:
            rev_dfs(u)
            scc_count += 1
    return scc_count

print(kosaraju(5, [[1,0],[0,2],[2,1],[0,3],[3,4]]))
print(kosaraju(4, [[0,1],[1,2],[2,3]]))

File: DSA/VaniJain_590012020/Day89Question3.c
#include <stdio.h>
#include <stdlib.h>
#define MAX 100
int V, E;
int graph[MAX][MAX];       
int transpose[MAX][MAX];  
int visited[MAX];
int stack[MAX], top = -1;
void dfs1(int v) 
{
    visited[v] = 1;
    for (int i = 0; i < V; i++) 
    {
        if (graph[v][i] && !visited[i]) 
        {
            dfs1(i);
        }
    }
    stack[++top] = v;
}
void dfs2(int v) 
{
    visited[v] = 1;
    printf("%d ", v);
    for (int i = 0; i < V; i++) 
    {
        if (transpose[v][i] && !visited[i]) 
        {
            dfs2(i);
        }
    }
}
void kosaraju() 
{
    for (int i = 0; i < V; i++)
        if (!visited[i])
            dfs1(i);
    for (int i = 0; i < V; i++)
        for (int j = 0; j < V; j++)
            transpose[i][j] = graph[j][i];
    for (int i = 0; i < V; i++) visited[i] = 0;
    int sccCount = 0;
    while (top >= 0) 
    {
        int v = stack[top--];
        if (!visited[v]) 
        {
            printf("SCC %d: ", ++sccCount);
            dfs2(v);
            printf("\n");
        }
    }
    printf("\nTotal Strongly Connected Components = %d\n", sccCount);
}
int main() 
{
    printf("Enter number of vertices: ");
    scanf("%d", &V);
    printf("Enter number of edges: ");
    scanf("%d", &E);
    for (int i = 0; i < V; i++)
        for (int j = 0; j < V; j++)
            graph[i][j] = 0;
    printf("Enter edges (u v) [0-indexed]:\n");
    for (int i = 0; i < E; i++) 
    {
        int u, v;
        scanf("%d %d", &u, &v);
        graph[u][v] = 1; 
    }
    kosaraju();
    return 0;
}
